File: lib/src/model/tor.dart
import 'dart:async';

import 'package:flutter/widgets.dart';
import 'package:path_provider/path_provider.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../rust/api/tor.dart' as core;

/// How far along the bundled Tor is, as the app says it.
///
/// The app's own type rather than the generated one, because it has a state
/// the core does not: **off**. The core can only describe a Tor that is
/// running, and "not running" is the state this spends most of its life in.
@immutable
class TorProgress {
  const TorProgress({
    required this.ready,
    required this.progress,
    required this.summary,
    this.blocked,
  });

  /// Not running at all.
  static const off = TorProgress(ready: false, progress: 0, summary: 'Off');

  /// Connections through it will now succeed.
  final bool ready;

  /// 0.0 to 1.0.
  final double progress;

  /// One line, from Tor, in its words.
  final String summary;

  /// Why it is stuck, when Tor knows — a censored network, no route out, a
  /// clock too far off for anything to verify. The only part of this a user
  /// can act on, so it is the part shown in full.
  final String? blocked;
}

/// The bundled Tor — beta.
///
/// Owns one question: is Tor running, and where. Everything about *using* it
/// is the proxy model's business, and deliberately so — this ends at a port
/// number, and [ProxySettings] decides whether anything goes through it.
///
/// That split is the whole reason bundling Tor was a small change. The app has
/// been able to reach Tor through a SOCKS5 proxy since the proxy setting
/// existed; what was missing was Tor. So this supplies one, on loopback,
/// speaking the protocol the app already speaks, and nothing downstream had to
/// learn a new way to connect.
class TorSettings extends ChangeNotifier {
  TorSettings._(this._prefs);

  static const _kEnabled = 'tor.enabled';

  final SharedPreferences? _prefs;

  bool _enabled = false;
  int? _port;
  TorProgress _progress = TorProgress.off;
  String? _failure;
  StreamSubscription<core.TorStatus>? _following;

  /// What the user asked for. Off by default, like everything else here — a
  /// few hundred megabytes of directory consensus is not something to fetch
  /// on behalf of someone who never asked.
  bool get enabled => _enabled;

  /// The loopback port Tor is answering SOCKS5 on, or null if it is not up.
  int? get port => _port;

  /// Running, whatever state the bootstrap is in.
  bool get running => _port != null;

  TorProgress get progress => _progress;

  /// Why the last attempt to start failed, if it did. Cleared by a successful
  /// start, so a stale message cannot outlive the problem.
  String? get failure => _failure;

  /// Read the preference. Does **not** start Tor: at the point settings are
  /// loaded the native core is not up yet, and starting it needs the core.
  static Future<TorSettings> load() async {
    SharedPreferences? prefs;
    try {
      prefs = await SharedPreferences.getInstance();
    } catch (e) {
      debugPrint('Tor preference unavailable, starting off: $e');
    }
    final tor = TorSettings._(prefs);
    tor._enabled = prefs?.getBool(_kEnabled) ?? false;
    return tor;
  }

  /// Start Tor if the user left it on. Called once, after the core is up and
  /// before anything auto-connects.
  ///
  /// Awaiting this matters: it returns as soon as the port is bound, which is
  /// long before Tor can carry anything, and a network dialled in that gap
  /// waits for the bootstrap rather than failing. What must not happen is a
  /// connection resolving its proxy before the port exists, because that is
  /// the one case that would look like "Tor is on" and travel direct.
  Future<void> startIfEnabled() async {
    if (!_enabled) return;
    await _start();
  }

  /// Turn it on or off. Writes through immediately.
  Future<void> setEnabled(bool value) async {
    if (value == _enabled && (!value || running)) return;
    _enabled = value;
    await _prefs?.setBool(_kEnabled, value);
    notifyListeners();
    if (value) {
      await _start();
    } else {
      await _stop();
    }
  }

  Future<void> _start() async {
    _failure = null;
    _progress = const TorProgress(
      ready: false,
      progress: 0,
      summary: 'Starting Tor',
    );
    notifyListeners();

    try {
      // Tor's own directory, inside the app's private storage. The consensus
      // it caches is public, but the guard list is not: it is a record of the
      // handful of relays this machine enters the network through, which is
      // exactly the thing worth keeping to itself.
      final base = await getApplicationSupportDirectory();
      _port = await core.torStart(dataDir: base.path);
    } catch (e) {
      _port = null;
      _failure = '$e';
      _progress = TorProgress.off;
      // Left switched on rather than silently flipped off: the user asked for
      // Tor, this failed to give it to them, and a switch that turns itself
      // off would hide that. Nothing connects through it while `port` is null.
      notifyListeners();
      return;
    }

    await _following?.cancel();
    _following = core.torStatusStream().listen((status) {
      _progress = TorProgress(
        ready: status.ready,
        progress: status.progress,
        summary: status.summary,
        blocked: status.blocked,
      );
      notifyListeners();
    }, onError: (Object e) => debugPrint('Tor status stream ended: $e'));
    notifyListeners();
  }

  Future<void> _stop() async {
    await _following?.cancel();
    _following = null;
    _port = null;
    _progress = TorProgress.off;
    notifyListeners();
    try {
      await core.torStop();
    } catch (e) {
      debugPrint('Tor did not stop cleanly: $e');
    }
  }

  @override
  void dispose() {
    unawaited(_following?.cancel());
    super.dispose();
  }
}

/// Makes [TorSettings] available to the widget tree.
class TorScope extends InheritedNotifier<TorSettings> {
  const TorScope({super.key, required TorSettings tor, required super.child})
    : super(notifier: tor);

  static TorSettings of(BuildContext context) {
    final scope = context.dependOnInheritedWidgetOfExactType<TorScope>();
    assert(scope?.notifier != null, 'No TorScope above this widget');
    return scope!.notifier!;
  }
}

File: irc-core/ddirc-bridge/src/api/tor.rs
//! Turning the bundled Tor on and off, and saying how far along it is.
//!
//! One at a time, and owned here: Dart gets a port number, not a handle. There
//! is exactly one embedded Tor per process because there is exactly one reason
//! to have it — a second would mean a second directory cache, a second set of
//! guards, and two answers to "is Tor on".
//!
//! What Dart does with the port is point the existing proxy setting at it. See
//! `ddirc-tor` for why bundling Tor deliberately did not add a new way for a
//! connection to leave the app.

use std::sync::{Mutex, OnceLock};

use ddirc_tor::TorService;

use crate::api::client::runtime;
use crate::frb_generated::StreamSink;

/// How far along Tor is, for the settings screen.
pub struct TorStatus {
    /// Connections through the proxy will now succeed. Before this they
    /// *wait*, rather than failing — so a network dialled early connects late
    /// rather than erroring.
    pub ready: bool,

    /// 0.0 to 1.0.
    pub progress: f32,

    /// One line, in Arti's words.
    pub summary: String,

    /// Why it is stuck, when Arti knows — a censored network, no route out, a
    /// clock too far off for anything to verify. The only part of this the
    /// user can act on, so it is the part shown in full.
    pub blocked: Option<String>,
}

impl From<ddirc_tor::TorStatus> for TorStatus {
    fn from(s: ddirc_tor::TorStatus) -> Self {
        TorStatus {
            ready: s.ready,
            progress: s.progress,
            summary: s.summary,
            blocked: s.blocked,
        }
    }
}

fn service() -> &'static Mutex<Option<TorService>> {
    static SERVICE: OnceLock<Mutex<Option<TorService>>> = OnceLock::new();
    SERVICE.get_or_init(|| Mutex::new(None))
}

/// Start the bundled Tor and return the loopback port it answers SOCKS5 on.
///
/// Returns as soon as there is a port, which is well before Tor can carry
/// anything: bootstrapping needs a directory consensus off the network, and
/// how long that takes depends on the network. Follow [`tor_status_stream`]
/// for the rest.
///
/// Starting when it is already running returns the port it is already on
/// rather than an error, so a settings screen rebuilt at the wrong moment
/// cannot end up with two.
///
/// `data_dir` is where Tor keeps its state and its directory cache. It must be
/// the app's own private storage: the cache is not secret, but the guard list
/// is a record of how this machine reaches the network.
pub fn tor_start(data_dir: String) -> Result<u16, String> {
    let mut guard = service().lock().unwrap_or_else(|e| e.into_inner());
    if let Some(running) = guard.as_ref() {
        return Ok(running.port());
    }
    let started = runtime()
        .block_on(TorService::start(std::path::Path::new(&data_dir)))
        .map_err(|e| e.to_string())?;
    let port = started.port();
    *guard = Some(started);
    Ok(port)
}

/// Stop it. Idempotent, so a switch turned off twice is not an error.
///
/// Connections currently routed through it are cut. That is deliberate: after
/// turning Tor off, nothing should still be connected by way of it, and
/// nothing should quietly fall back to a direct route it was never given
/// permission to take.
pub fn tor_stop() {
    let taken = service().lock().unwrap_or_else(|e| e.into_inner()).take();
    if let Some(service) = taken {
        runtime().block_on(service.stop());
    }
}

/// The port Tor is on, or null if it is not running.
///
/// For a UI coming back to a screen rather than one that just pressed the
/// switch — the answer has to survive the widget being rebuilt.
pub fn tor_port() -> Option<u16> {
    service()
        .lock()
        .unwrap_or_else(|e| e.into_inner())
        .as_ref()
        .map(|s| s.port())
}

/// Follow the bootstrap. In Dart this is a `Stream`.
///
/// The current status is delivered first, so a screen that subscribes late
/// still knows where things stand rather than waiting for the next change.
/// The stream ends when Tor stops.
pub fn tor_status_stream(sink: StreamSink<TorStatus>) -> Result<(), String> {
    let mut watch = {
        let guard = service().lock().unwrap_or_else(|e| e.into_inner());
        let service = guard.as_ref().ok_or("Tor is not running")?;
        service.watch()
    };

    runtime().spawn(async move {
        // `borrow_and_update` rather than `borrow`, so the first `changed`
        // reports the next change rather than immediately returning the value
        // that was just sent.
        let current = watch.borrow_and_update().clone();
        if sink.add(TorStatus::from(current)).is_err() {
            return;
        }
        while watch.changed().await.is_ok() {
            let status = watch.borrow_and_update().clone();
            if sink.add(TorStatus::from(status)).is_err() {
                return;
            }
        }
    });
    Ok(())
}
